enterprise/backup/src/main/java/org/neo4j/backup/OnlineBackupCommandBuilder.java

```diff
@@ -0,0 +1,361 @@
+        return this;
+    }
+
+    /**
+     * Perform a backup with the parameters set using the builder
+     * @param targetLocation target location of backup (NOTE: uses the parent directory as well)
+     * @return true if backup was successful
+     */
+    public boolean backup( String targetLocation ) throws CommandFailed, IncorrectUsage
+    {
+        return backup( relativeFileFromString( targetLocation ) );
+    }
+
+    /**
+     * Perform a backup with the parameters set using the builder
+     * @param targetLocation target location of backup (NOTE: uses the parent directory as well)
+     * @return true if backup was successful
+     */
+    public boolean backup( File targetLocation ) throws CommandFailed, IncorrectUsage
+    {
+        new OnlineBackupCommandProvider()
+                .create( neo4jHomeFromTarget( targetLocation ),
+                        configDirFromTarget( targetLocation ),
+                        resolveOutsideWorld() )
+                .execute( resolveArgs( targetLocation ) );
+        return true;
+    }
+
+    private File relativeFileFromString( String targetLocation )
+    {
+        return new File( "." ).toPath().resolve( targetLocation ).toFile();
+    }
+
+    /**
+     * This is a helper method when debugging tests where in some cases the file is a relative string
+     * @param targetLocation relative backup location as a string
+     * @return the arguments that would be passed to the backup command
+     */
+    public String[] resolveArgs( String targetLocation )
+    {
+        return resolveArgs( relativeFileFromString( targetLocation ) );
+    }
+
+    public String[] resolveArgs( File targetLocation )
+    {
+        return args(
+                argBackupName( targetLocation ),
+                argBackupLocation( targetLocation ),
+                argFrom(),
+                argFallbackToFull(),
+                argTimeout(),
+                argCheckConsistency(),
+                argReportDir(),
+                argAdditionalConf( targetLocation ),
+                argCcGraph(),
+                argCcIndexes(),
+                argCcLabel(),
+                argCcOwners() );
+    }
+
+    private OutsideWorld resolveOutsideWorld()
+    {
+        Optional<OutputStream> output = Optional.ofNullable( this.output );
+        return new ParametrisedOutsideWorld( System.console(), output.orElse( System.out ), output.orElse( System.err ), new DefaultFileSystemAbstraction() );
+    }
+
+    /**
+     * Client handles the ports and hosts automatically, so no necessary need to specify
+     * @return command line parameter for specifying the backup address
+     */
+    private String argFrom()
+    {
+        if ( host == null && port == null )
+        {
+            return "";
+        }
+        String address = String.join( ":",
+                Optional.ofNullable( host ).orElse( "" ),
+                Optional.ofNullable( port ).map( port -> Integer.toString( port ) ).orElse( "" ) );
```

Comment:
Might be cleaner to store host, port (and other fields) as Optional rather than wrap them here. Intellij will complain, but it's wrong :)

Reply:
Removed file (included in separate branch)